### PR TITLE
Levenshtein improvements

### LINKED ISSUE #224: Levenshtein distance calculation is slow
Bug report sent in by a WatchDog user:

> I'm afraid I've got a new one though... The method com.cedarsoftware.util.StringUtilities.levenshteinDistance()  is taking far far more time than is reasonable, and is regularly pushing all cores to 100%, generally killing my general development experience. I've restarted IntelliJ and the machine itself, and cleared intelliJ's caches, but no effect. 
>
> Profiling IntelliJ for about 10 seconds of moderate file manipulation shows 250+ seconds of cpu time spent in that method. I don't have 25 cores, so I'm not sure what that means, perhaps the sampler is playing fast and loose with timings.
>
> It is called by nl.tudelft.watchdog.core.logic.interval.intervaltypes.TypingInterval$TypingIntervalCloserBase.run(), which itself is called by Thread.run(), and usually pegs all cores for about a minute or two at least before calming down again. 
>
> Attached is the CPU profiling session from jvisualvm, in the hopes that it sheds some light on the issue.  I've been hunting this bug in my workspace for much of the afternoon, and unfortunately need to disable this plugin until this issue is resolved.
>
> Sorry to trade one fix for another bug report, but I'm certain that we're all committed to less buggy software, and the only way to get there is from more bug reports!

### PR DESCRIPTION
**Pull request by nspruit**
This PR contains an implementation of the three suggestions in #224. Therefore, this PR should close #224

### CODE REVIEW

File: WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/interval/intervaltypes/TypingInterval.java [hunk 1]
```diff
@@ -32,9 +38,23 @@ public void run() {
 				String startingContent = getDocument().getContent();
 				String endingContent = endingDocument.getContent();
 				if (startingContent != null && endingContent != null) {
-					editDistance = new JsonifiedLong(
-							StringUtilities.levenshteinDistance(
-									startingContent, endingContent));
+					int startLength = startingContent.length();
+					int endLength = endingContent.length();
+					charLengthDiffBetweenStartAndEndDoc = Math.abs(startLength - endLength);
```
Comment (Inventitech) [inline WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/interval/intervaltypes/TypingInterval.java]: Shorten to charLengthDiff ?

File: WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/interval/intervaltypes/TypingInterval.java [hunk 2]
```diff
@@ -71,6 +106,13 @@ public IntervalType getType() {
 	public void setEndingDocument(Document endingDocument) {
 		this.endingDocument = endingDocument;
 	}
+	
+	/** Increases the number of characters modified in this interval. */
+	public void increaseModCountWith(long nrOfModifiedChars) {
```
Comment (Inventitech) [inline WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/interval/intervaltypes/TypingInterval.java]: simplify to modifiedChars (it is a long anyway)?

File: WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/ui/EventManagerBase.java [hunk 3]
```diff
@@ -147,6 +147,11 @@ public void update(WatchDogEvent event, Date forcedDate) {
 						EventType.START_EDIT));
 				break;
 			}
+			
+			//increase modCount for the current typing interval
```
Comment (Inventitech) [inline WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/ui/EventManagerBase.java]: Please begin comments with a space and upper case letter.

File: WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/ui/EventManagerBase.java [hunk 4]
```diff
@@ -147,6 +147,11 @@ public void update(WatchDogEvent event, Date forcedDate) {
...
+			EditorEvent edEvent = (EditorEvent) event;
+			TypingInterval currInterval = (TypingInterval) editorInterval;
```
Comment (Inventitech) [inline WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/ui/EventManagerBase.java]: Typo: currInterval? Rename to simply typingInterval or similar?

File: WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/ui/EventManagerBase.java [hunk 5]
```diff
@@ -147,6 +147,11 @@ public void update(WatchDogEvent event, Date forcedDate) {
 						EventType.START_EDIT));
 				break;
 			}
+			
+			//increase modCount for the current typing interval
+			EditorEvent edEvent = (EditorEvent) event;
```
Comment (Inventitech) [inline WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/ui/EventManagerBase.java]: Please use a different name, like for example editorInterval

File: WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EditorListener.java [hunk 6]
```diff
@@ -53,8 +53,19 @@ private void listenToDocumentChanges() throws IllegalArgumentException {
...
-						EventType.SUBSEQUENT_EDIT));
+				/**
+				 * Three events exist that can influence the Levenshtein distance:
+				 * 1. Addition. In this case length=0 and text>0, therefore max(length,text)=text=Levenshtein distance.
+				 * 2. Removal. In this case length>0 and text=0, therefore max(length,text)=length=Levenshtein distance.
+				 * 3. Modification. In this case length>0 and text>0, therefore max(length,text)>=Levenshtein distance.
+				 * 
+				 * So, in general it holds that modCount >= Levenshtein distance.
+				 */
+				int modCount = Math.max(event.getLength(),
```
Comment (Inventitech) [inline WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EditorListener.java]: Can event.getLength() or even more so, event.getText().length() become null here? Do we need to protect against this?

**Comment by Inventitech** [comment]
A few minor comments.

File: WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EditorListener.java (hunk 6 above)
Comment (Inventitech) [inline WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EditorListener.java]: This looks like a JavaDoc comment (double asterisk), but should be a regular multi-line comment.

**Comment by nspruit** [comment]
Thanks for the feedback! I've processed it in the latest commit

File: WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/interval/intervaltypes/TypingInterval.java [hunk 7]
```diff
@@ -56,10 +76,25 @@ public void run() {
...
+	 * this interval.
+	 */	
+	@SerializedName("modCount")
+	private int modCount;
+	
+	/**
+	 * The difference in the number of characters between the starting and ending
+	 * document.
+	 */
+	@SerializedName("startEndDiff")
```
Comment (Inventitech) [inline WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/interval/intervaltypes/TypingInterval.java]: Do you have a reason to have the serializedName differ from the variable name?

File: WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/interval/intervaltypes/TypingInterval.java [hunk 8]
```diff
@@ -56,10 +76,25 @@ public void run() {
 	 */
 	@SerializedName("diff")
 	JsonifiedLong editDistance;
+	
+	/**
+	 * The number of characters that have been added, removed or modified during
+	 * this interval.
+	 */	
+	@SerializedName("modCount")
```
Comment (Inventitech) [inline WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/interval/intervaltypes/TypingInterval.java]: Could you postfix "Diff" to make it apparent that diff, modCountDiff and startEndDiff belong together?

File: WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/interval/intervaltypes/TypingInterval.java [hunk 9]
```diff
@@ -32,9 +38,23 @@ public void run() {
...
+					
+					long lengthProduct = startLength*endLength;
+					if(startLength == 0) {
+						lengthProduct = endLength;
+					} 
+					else if (endLength == 0) {
+						lengthProduct = startLength;
+					}
+					
+					//only calculate Levenshtein when its impact on the usability is acceptable
```
Comment (Inventitech) [inline WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/interval/intervaltypes/TypingInterval.java]: Comment should begin with capital letter and space after //

File: WatchDogEclipsePlugin/WatchDogUnitTests/src/nl/tudelft/watchdog/logic/interval/JsonConverterTest.java [hunk 10]
```diff
@@ -65,7 +65,7 @@ public void testJsonTypingIntervalMissingDocumentRepresentation() {
 		ArrayList<IntervalBase> intervals = createSampleIntervals(interval);
 
 		assertEquals(
-				"[{\"doc\":{\"pn\":\"f6f4da8d93e88a08220e03b7810451d3ba540a34\",\"fn\":\"90a8834de76326869f3e703cd61513081ad73d3c\",\"sloc\":1,\"dt\":\"pr\"},\"it\":\"ty\",\"ts\":1,\"te\":2,\"ss\":\"\","
+				"[{\"modCount\":0,\"startEndDiff\":0,\"doc\":{\"pn\":\"f6f4da8d93e88a08220e03b7810451d3ba540a34\",\"fn\":\"90a8834de76326869f3e703cd61513081ad73d3c\",\"sloc\":1,\"dt\":\"pr\"},\"it\":\"ty\",\"ts\":1,\"te\":2,\"ss\":\"\","
```
Comment (Inventitech) [inline WatchDogEclipsePlugin/WatchDogUnitTests/src/nl/tudelft/watchdog/logic/interval/JsonConverterTest.java]: Would it be possible to add a few testcases to showcase the newly added features? (At least the startEndDiff.)

File: WatchDogIntelliJPlugin/WatchDog/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorListener.java [hunk 11]
```diff
@@ -43,7 +43,24 @@ public void beforeDocumentChange(DocumentEvent event) {
 
             @Override
 			public void documentChanged(DocumentEvent event) {
-                eventManager.update(new EditorEvent(editor, EventType.SUBSEQUENT_EDIT));
+				/**
+				 * Three events exist that can influence the Levenshtein distance:
```
Comment (Inventitech) [inline WatchDogIntelliJPlugin/WatchDog/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorListener.java]: JavaDoc-style comment inside method?

File: WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/interval/intervaltypes/TypingInterval.java (hunk 7 above)
Comment (nspruit) [inline WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/interval/intervaltypes/TypingInterval.java]: Not anymore. This serialized named belonged to the previous variable name, which was much longer. I'll fix this

File: WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/interval/intervaltypes/TypingInterval.java (hunk 8 above)
Comment (nspruit) [inline WatchDogCore/WatchDog/src/nl/tudelft/watchdog/core/logic/interval/intervaltypes/TypingInterval.java]: Sure, will do

**Comment by Inventitech** [comment]
Excellent.

**Pull request merged**